src/neml2/models/Model.cxx

```diff
@@ -199,14 +225,132 @@ Model::zero_output()
     submodel->zero_output();
 }
 
+Model::TraceSchema
+Model::compute_trace_schema() const
+{
+  std::vector<Size> batch_dims;
+  for (auto && [name, var] : input_variables())
+    batch_dims.push_back(var.batch_dim());
+  for (auto && [name, param] : host<ParameterStore>()->named_parameters())
+    batch_dims.push_back(Tensor(param).batch_dim());
+
+  return TraceSchema{batch_dims};
+}
+
+std::size_t
+Model::forward_operator_index(bool out, bool dout, bool d2out) const
+{
+  return (out ? 4 : 0) + (dout ? 2 : 0) + (d2out ? 1 : 0);
+}
+
+void
+Model::forward(bool out, bool dout, bool d2out)
+{
+  torch::InferenceMode mode_guard(_production && !torch::jit::tracer::isTracing());
+
+  if (dout || d2out)
+    enable_AD();
+
+  set_value(out || AD_need_value(dout, d2out), dout, d2out);
+
+  if (dout || d2out)
+    extract_AD_derivatives(dout, d2out);
+
+  return;
+}
+
+void
+Model::forward_maybe_jit(bool out, bool dout, bool d2out)
+{
+  if (!is_jit_enabled() || torch::jit::tracer::isTracing())
+  {
+    forward(out, dout, d2out);
+    return;
+  }
+
+  auto & traced_functions =
+      currently_solving_nonlinear_system() ? _traced_functions_nl_sys : _traced_functions;
+
+  const auto forward_op_idx = forward_operator_index(out, dout, d2out);
+  const auto new_schema = compute_trace_schema();
+  auto traced_schema_and_function = traced_functions[forward_op_idx].find(new_schema);
+
+  if (traced_schema_and_function != traced_functions[forward_op_idx].end())
+  {
+    auto & [trace_schema, traced_function] = *traced_schema_and_function;
+    torch::InferenceMode mode_guard(_production);
+    auto stack = collect_input_stack();
+    traced_function->run(stack);
+    assign_output_stack(stack, out, dout, d2out);
+  }
+  else
+  {
+    auto forward_wrap = [&](torch::jit::Stack inputs) -> torch::jit::Stack
+    {
+      assign_input_stack(inputs);
+      forward(out, dout, d2out);
+      return collect_output_stack(out, dout, d2out);
+    };
+    auto trace = std::get<0>(torch::jit::tracer::trace(
+        collect_input_stack(),
+        forward_wrap,
+        [this](const torch::Tensor & var) { return variable_name_lookup(var); },
+        /*strict=*/false,
+        /*force_outplace=*/false));
+    auto new_function =
+        std::make_unique<torch::jit::GraphFunction>(name() + ".forward",
+                                                    trace->graph,
+                                                    /*function_creator=*/nullptr,
+                                                    torch::jit::ExecutorExecutionMode::PROFILING);
+    traced_functions[forward_op_idx].emplace(new_schema, std::move(new_function));
+
+    // Rerun this method -- this time using the jitted graph (without tracing)
+    forward_maybe_jit(out, dout, d2out);
```

Comment:
Is it necessary to rerun?  It seems like it is, but surely the results are calculated along with the trace, or else how would they be checking?  I can't find anything in the API to give you the (checked) results along with the traced code.

Reply:
Me neither...